### ISSUE TITLE
[TypeConverter] attribute bug

### ISSUE THREAD

**Issue by richard-browne** [JoshClose/CsvHelper#1063]
Take the following example:

```cs
        struct MyStruct
        {
        }

        class MyCsv
        {
            public string Col1 { get; set; }
            [TypeConverter(typeof(MyStructConverter))]
            public MyStruct Col2 { get; set; }
        }

        class MyStructConverter : CsvHelper.TypeConversion.DefaultTypeConverter
        {
        }

        static void WriteCsv(string filename, IList<MyCsv> lines)
        {
            using (var csvWriter = new CsvWriter(new StreamWriter(filename)))
            {
                csvWriter.WriteRecords(lines);
            }
        }

```

This doesn't work. The custom MyStructConverter is not used. Instead Col2 is removed from the output altogether.

I think the problem is in ClassMap.AutoMapMembers(). It doesn't look to see if a member has a [TypeConverter] attribute before calling configuration.TypeConverterCache.GetConverter().

**Comment by JoshClose** [comment]
You'll need to give me a failing example. This seems to work.

```cs
void Main()
{
	var records = new List<Test>
	{
		new Test { Id = 1, Name = "one" },
		new Test { Id = 2, Name = "two" },
	};
	
	using (var writer = new StringWriter())
	using (var csv = new CsvWriter(writer))
	{
		csv.WriteRecords(records);
		
		writer.ToString().Dump();
	}
}

public struct Test
{
	public int Id { get; set; }
	[CsvHelper.Configuration.Attributes.TypeConverter(typeof(MyStructConverter))]
	public string Name { get; set; }
}

public class MyStructConverter : DefaultTypeConverter
{
	public override string ConvertToString(object value, IWriterRow row, MemberMapData memberMapData)
	{
		return "test";
	}
}
```

**Comment by richard-browne** [comment]
Man I did provide a failing example. The reason your code works is because Test.Name is a string. Try the following modification. The second column is removed from the output:

```
static void Main()
{
    var records = new List<Test>
    {
        new Test { Id = 1, Name = new AStruct() },
        new Test { Id = 2, Name = new AStruct() },
    };

    using (var writer = new StringWriter())
    using (var csv = new CsvWriter(writer))
    {
        csv.WriteRecords(records);

        Console.WriteLine(writer.ToString());
    }
}

public struct AStruct { }

public struct Test
{
    public int Id { get; set; }
    [CsvHelper.Configuration.Attributes.TypeConverter(typeof(MyStructConverter))]
    public AStruct Name { get; set; }
}

public class MyStructConverter : DefaultTypeConverter
{
    public override string ConvertToString(object value, IWriterRow row, MemberMapData memberMapData)
    {
        return "test";
    }
}

```

**Comment by JoshClose** [comment]
Oh sorry, I read your example wrong.

Things don't work that way. If you have a class or a struct as a property, the properties/fields of that class/struct are looked at individually.

```cs
void Main()
{
	var records = new List<Test>
	{
		new Test { Id = 1, TheStruct = new MyStruct { Name = "one" } },
		new Test { Id = 2, TheStruct = new MyStruct { Name = "two" } },
	};
	
	using (var writer = new StringWriter())
	using (var csv = new CsvWriter(writer))
	{
		csv.WriteRecords(records);
		
		writer.ToString().Dump();
	}
}

public class Test
{
	public int Id { get; set; }
	public MyStruct TheStruct { get; set; }
}

public struct MyStruct
{
	[TypeConverter(typeof(MyConverter))]
	public string Name { get; set; }
}

public class MyConverter : DefaultTypeConverter
{
	public override string ConvertToString(object value, IWriterRow row, MemberMapData memberMapData)
	{
		return $"{value} blah";
	}
}
```

![image](https://user-images.githubusercontent.com/114971/42228148-d2a96a08-7ea8-11e8-974c-a5ef2884c7ed.png)

**Comment by richard-browne** [comment]
Thanks for the reply. If you say "things don't work that way" then I can't argue since you are the author. But allow me to disagree anyway? :-)

If I create a CsvHelper.Configuration configuration object and call:

configuration.TypeConverterCache.AddConverter<MyStruct>(new MyConverter());

Everything works fine - as I expect. So are you saying the above works, but the equivalent [TypeConverter(typeof(MyConverter))] attribute doesn't work.. by design? I don't think you do really mean that. I think the attribute and the 'AddConverter' calls are supposed to be equivalent.

In my case I have a struct that has no public properties or fields. Just like System.DateTime.

I think this is a bug. In ClassMap.AutoMapMembers(). we need to check for the attribute before deciding DefaultTypeConverter is in play (line 286 of ClassMap.cs in master as I type).

**Comment by richard-browne** [comment]
Also.. if we could agree my sample should work as *I* want it to, I'd be most happy to submit a pull request with some unit tests.

**Comment by JoshClose** [comment]
It works if you register it on a mapping also. This is definitely a bug.

**Comment by JoshClose** [comment]
Fixed and in version 9.2.1 on nuget.
